grades/api_test.py

```diff
@@ -0,0 +1,232 @@
+                user=cls.user, course_run=course_run) for course_run in all_course_runs
+        }
+
+        cls.certificates = {
+            course_run.edx_course_key: CachedCertificateFactory.create(
+                user=cls.user, course_run=course_run) for course_run in (cls.course_run2, cls.course_run4)
+        }
+
+        cls.user_edx_data = CachedEdxUserData(cls.user)
+
+    def test_compute_grade_for_fa(self):
+        """
+        Tests for _compute_grade_for_fa function
+        """
+        run1_data = self.user_edx_data.get_run_data(self.course_run1.edx_course_key)
+        run2_data = self.user_edx_data.get_run_data(self.course_run2.edx_course_key)
+
+        grade1_from_cur_grade = api._compute_grade_for_fa(run1_data)
+        grade2_from_cert = api._compute_grade_for_fa(run2_data)
+
+        assert isinstance(grade1_from_cur_grade, api.UserFinalGrade)
+        assert isinstance(grade2_from_cert, api.UserFinalGrade)
+
+        assert grade1_from_cur_grade.passed == self.current_grades.get(
+            self.course_run1.edx_course_key).data.get('passed')
+        assert grade1_from_cur_grade.grade == self.current_grades.get(
+            self.course_run1.edx_course_key).data.get('percent')
+
+        assert grade2_from_cert.passed is True
+        assert grade2_from_cert.grade == self.certificates.get(
+            self.course_run2.edx_course_key).data.get('grade')
+
+    def test_compute_grade_for_non_fa(self):
+        """
+        Tests for _compute_grade_for_non_fa function
+        """
+        run3_data = self.user_edx_data.get_run_data(self.course_run3.edx_course_key)
+        run4_data = self.user_edx_data.get_run_data(self.course_run4.edx_course_key)
+
+        grade3 = api._compute_grade_for_non_fa(run3_data)
+        grade4 = api._compute_grade_for_non_fa(run4_data)
+
+        assert isinstance(grade3, api.UserFinalGrade)
+        assert isinstance(grade4, api.UserFinalGrade)
+
+        assert grade3.passed is False
+        assert grade3.grade == self.current_grades.get(
+            self.course_run3.edx_course_key).data.get('percent')
+        assert grade4.passed is True
+        assert grade4.grade == self.certificates.get(
+            self.course_run4.edx_course_key).data.get('grade')
+
+    def test_get_compute_func(self):
+        """
+        tests for _get_compute_func function
+        """
+        assert api._get_compute_func(self.course_run1) == api._compute_grade_for_fa
+        assert api._get_compute_func(self.course_run3) == api._compute_grade_for_non_fa
+
+    @patch('grades.api._compute_grade_for_non_fa', autospec=True)
+    @patch('grades.api._compute_grade_for_fa', autospec=True)
+    def test_get_final_grade(self, fa, non_fa):
+        """
+        tests for get_final_grade function
+        """
+        cruns = (self.course_run1, self.course_run3,)
+
+        for crun in cruns:
+            api.get_final_grade(self.user, crun)
```

Comment:
Does this mutate any state? It seems like it's just calculating the grade, not storing it, though I haven't read through it thoroughly yet